engine-nextgen/src/main/kotlin/graphql/nadel/enginekt/transform/result/json/JsonNodes.kt

```diff
@@ -0,0 +1,148 @@
+package graphql.nadel.enginekt.transform.result.json
+
+import graphql.nadel.NadelExecutionHints
+import graphql.nadel.enginekt.transform.query.NadelQueryPath
+import graphql.nadel.enginekt.util.AnyList
+import graphql.nadel.enginekt.util.AnyMap
+import graphql.nadel.enginekt.util.JsonMap
+import graphql.nadel.enginekt.util.foldWhileNotNull
+import java.util.Collections
+
+/**
+ * Utility class to extract data out of the given
+ */
+class JsonNodes(
+    private val data: JsonMap,
+    private val executionFlags: NadelExecutionHints,
+) {
+    private val nodes = Collections.synchronizedMap(
```

Comment:
So even though we were traversing the tree specifically according to the input path, if there were a _lot_ of arrays, we would almost traverse the entire tree multiple times e.g.

```json
{
  "issues": [
    {
      "id": "hello",
      "desc": "test"
    },
    {
      "id": "hello",
      "desc": "test"
    },
    {
      "id": "hello",
      "desc": "test"
    },
    {
      "id": "hello",
      "desc": "test"
    }
  ]
}
```

So to access `id` you have to iterate through 4 objects in the array, and to access `desc` it's the same. Now imagine this 10 objects/arrays deep e.g. `myActivities/workedOn/nodes/contributors/profile/{accountId, name, picture}` That's the real performance gain

Reply:
Double checking that you want a synchronized map here

`Collections.synchronizedMap`

So the `JsonNodes` code can be called from multiple threads yeah?

Reply:
Yup, it's behind a FF but we have a thing to parallelize the `getResultInstructions` calls which call this.